gds_write/def2stream.py

```diff
@@ -0,0 +1,208 @@
+tech.load(args.tech_file)
+layoutOptions = tech.load_layout_options
+if len(args.layer_map) > 0:
+  layoutOptions.lefdef_config.map_file = args.layer_map
+
+# Load def file
+main_layout = db.Layout()
+print("[INFO] Reporting cells prior to loading DEF ...")
+for i in main_layout.each_cell():
+  print("[INFO] '{0}'".format(i.name))
+
+print("[INFO] Reading DEF ...")
+main_layout.read(args.input_def, layoutOptions)
+
+#print("[INFO] Reporting cells after loading DEF ...")
+#for i in main_layout.each_cell():
+#  print("[INFO] '{0}'".format(i.name))
+
+# Clear cells
+top_cell_index = main_layout.cell(args.design_name).cell_index()
+
+# remove orphan cell BUT preserve cell with VIA_
+#  - KLayout is prepending VIA_ when reading DEF that instantiates LEF's via
+print("[INFO] Clearing cells...")
+for i in main_layout.each_cell():
+  if i.cell_index() != top_cell_index:
+    if not i.name.startswith("VIA_"):
+      i.clear()
+
+# Load in the gds to merge
+print("[INFO] Merging GDS/OAS files...")
+for fil in args.additional_gds:
+  print("\t{0}".format(fil))
+  main_layout.read(fil)
+
+# Copy the top level only to a new layout
+print("[INFO] Copying toplevel cell '{0}'".format(args.design_name))
+top_only_layout = db.Layout()
+top_only_layout.dbu = main_layout.dbu
+top = top_only_layout.create_cell(args.design_name)
+top.copy_tree(main_layout.cell(args.design_name))
+
+read_fills(top)
+
+print("[INFO] Checking for missing cell from GDS/OAS...")
+missing_cell = False
+regex = None
+if 'GDS_ALLOW_EMPTY' in os.environ:
+    print("[INFO] Found GDS_ALLOW_EMPTY variable.")
+    regex = os.getenv('GDS_ALLOW_EMPTY')
+for i in top_only_layout.each_cell():
+  if i.is_empty():
+    missing_cell = True
+    if regex is not None and re.match(regex, i.name):
+        print("[WARNING] LEF Cell '{0}' ignored. Matches GDS_ALLOW_EMPTY.".format(i.name))
+    else:
+        print("[ERROR] LEF Cell '{0}' has no matching GDS/OAS cell."
+              " Cell will be empty.".format(i.name))
+        errors += 1
+
+if not missing_cell:
+  print("[INFO] All LEF cells have matching GDS/OAS cells")
+
+print("[INFO] Checking for orphan cell in the final layout...")
+orphan_cell = False
+for i in top_only_layout.each_cell():
+  if i.name != args.design_name and i.parent_cells() == 0:
+    orphan_cell = True
+    print("[ERROR] Found orphan cell '{0}'".format(i.name))
+    errors += 1
+
+if not orphan_cell:
+  print("[INFO] No orphan cells")
+
+
+if args.seal:
+
+  top_cell = top_only_layout.top_cell()
+
+  print("[INFO] Reading seal GDS/OAS file...")
+  print("\t{0}".format(args.seal))
+  top_only_layout.read(args.seal)
+
+  for cell in top_only_layout.top_cells():
+    if cell != top_cell:
+      print("[INFO] Merging '{0}' as child of '{1}'".format(cell.name, top_cell.name))
+      top.insert(db.CellInstArray(cell.cell_index(), db.Trans()))
+
+# Write out the GDS
+print("[INFO] Writing out GDS/OAS '{0}'".format(args.out))
+top_only_layout.write(args.out)
+
+sys.exit(errors)
```

Comment:
should we use exception instead? so that can exit earlier in the flow?

Reply:
replaced with `raise LookupError()`